change libbot sha1 to use new libbot build system

patmarion opened the pull request.

patmarion commented:
Just got it working on Mac.  Need to test a lot more.  Opening pull request now to get some build server results.

patmarion commented:
Ok, I'm reasonably happy with this for now.  We'll see how the build servers do...

This branch simply updates the libbot sha1.  Currently it is set to a branch of my libbot fork, which has a pending pull request here:  https://github.com/RobotLocomotion/libbot/pull/12

Here is the new behavior:

  - libbot uses cmake ExternalProject instead of tobuild.txt
  - the libbot Makefile is now a copy of the "new style" Makefile I borrowed from other pods[1]
  - by default, WITH_BOT_VIS is set to OFF
  - in the default mode, WITH_BOT_VIS is OFF, and thus lcmgl doesn't have text/texture support
  - To enable bot2-vis, and thus build libbot completely as before:

    cd externals/libbot
    make options
    (using ccmake, set WITH_BOT_VIS to ON)
    make


[1] I noticed the new style Makefile does not search for a build directory, instead it assumes that BUILD_PREFIX has been passed in via the environment.  This isn't pods compliant anymore, but maybe we don't care?  It also doesn't implement make clean in the same way as before.

RussTedrake commented:
The makefile in the updated pods pushes the search for the build directory into the cmakefiles.txt .  It's at the top of the cmakelists in drake-distro and also in pods.cmake



> On Sep 29, 2015, at 5:31 PM, Pat Marion <notifications@github.com> wrote:
> 
> Ok, I'm reasonably happy with this for now.  We'll see how the build servers do...
> 
> This branch simply updates the libbot sha1. Currently it is set to a branch of my libbot fork, which has a pending pull request here: RobotLocomotion/libbot#12
> 
> Here is the new behavior:
> 
> libbot uses cmake ExternalProject instead of tobuild.txt
> the libbot Makefile is now a copy of the "new style" Makefile I borrowed from other pods[1]
> by default, WITH_BOT_VIS is set to OFF
> in the default mode, WITH_BOT_VIS is OFF, and thus lcmgl doesn't have text/texture support
> To enable bot2-vis, and thus build libbot completely as before:
> 
> cd externals/libbot
> make options
> (using ccmake, set WITH_BOT_VIS to ON)
> make
> 
> [1] I noticed the new style Makefile does not search for a build directory, instead it assumes that BUILD_PREFIX has been passed in via the environment. This isn't pods compliant anymore, but maybe we don't care? It also doesn't implement make clean in the same way as before.
> 
> —
> Reply to this email directly or view it on GitHub.
>

patmarion commented:
Ok, got it, added search logic for build directory to libbot's new CMakeLists.txt
I added logic to search for gdkx.h and use that to decide the default setting of WITH_BOT_VIS.

I was a little misleading when talking about text support.  The Director has never supported text, it's only available in drake_viewer.  It's possible we could bring text support into the Directory, but we need to link with and initialize the GLUT library.  So, WITH_BOT_VIS=OFF means you don't get drake_viewer at all.  WITH_BOT_VIS=ON means you get drake_viewer (and it has text support).

RussTedrake commented:
ok, i didn’t understand that.  i consider drake_viewer to be on the verge of extinction, and don’t think we should pollute any of the new stuff to support it unless there are places where directory really doesn’t work.

> On Sep 30, 2015, at 4:57 PM, Pat Marion <notifications@github.com> wrote:
> 
> Ok, got it, added search logic for build directory to libbot's new CMakeLists.txt
> I added logic to search for gdkx.h and use that to decide the default setting of WITH_BOT_VIS.
> 
> I was a little misleading when talking about text support. The Director has never supported text, it's only available in drake_viewer. It's possible we could bring text support into the Directory, but we need to link with and initialize the GLUT library. So, WITH_BOT_VIS=OFF means you don't get drake_viewer at all. WITH_BOT_VIS=ON means you get drake_viewer (and it has text support).
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/RobotLocomotion/drake/pull/1384#issuecomment-144437643>.
>

patmarion commented:
Build servers seem ok.  Note, in the last round drake001 should have compiled using WITH_BOT_VIS=ON because that machine has gtk+x11 installed via homebrew "hacks".

I think I'm done working on this for now, should be ready to go once.  Not sure if the update will go smoothly for people attempting incremental builds, but it might be fine.

patmarion commented:
This shouldn't be merged until after the libbot pull request though.

RussTedrake commented on CMakeLists.txt:
```diff
@@ -119,8 +119,8 @@ set(lcm_GIT_TAG 23876258aaf1c29a0a29915aa82c7b785214a2bf) # note: cmake branch
 set(lcm_IS_PUBLIC TRUE)
-set(libbot_GIT_REPOSITORY https://github.com/RobotLocomotion/libbot.git)
-set(libbot_GIT_TAG c328b73f3b48008d04468031e26f48468a7e3cc0)
+set(libbot_GIT_REPOSITORY https://github.com/patmarion/libbot.git)
```
ready to put this back to RobotLocomotion.  thanks Pat.

patmarion commented:
All set!  Now points to libbot master.

avalenzu commented:
Should we remove lcmgl text calls from the code base? Several of the
collision tests use it when run with visualize = true (it defaults to
false, so the build servers won't see it.

On Thu, Oct 1, 2015, 8:49 AM Pat Marion <notifications@github.com> wrote:

> All set! Now points to libbot master.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/1384#issuecomment-144717564>
> .
>

patmarion commented:
Btw, I think attempts to draw text with LCMGL in the director will result in warnings if WITH_BOT_VIS=OFF, or a crash if WITH_BOT_VIS=ON.  Someone could probably spend a couple hours at some point to determine the feasibility of bringing in lcmgl text support.  I think it requires linking to GLUT and calling the glut init function, and porting some texture allocation logic from bot2-vis into bot2-lcmgl-renderer.

rdeits commented:
After this change, I'm getting build errors in drake: 

```
[ 77%] Building CXX object systems/plants/viewer/CMakeFiles/drake_urdf_renderer.dir/drake_urdf_renderer.cpp.o
In file included from /Users/rdeits/locomotion/drake-distro/drake/systems/plants/viewer/drake_urdf_renderer.cpp:10:
In file included from /Users/rdeits/locomotion/drake-distro/build/include/bot_vis/bot_vis.h:6:
In file included from /Users/rdeits/locomotion/drake-distro/build/include/bot_vis/default_view_handler.h:8:
/Users/rdeits/locomotion/drake-distro/build/include/bot_vis/viewer.h:5:10: fatal error: 'gtk/gtk.h' file not
      found
#include <gtk/gtk.h>
         ^
1 error generated.
```

I tried many things:
* make clean (in drake and drake-distro)
* removing the libbot folder entirely and rebuilding it
* building libbot with WITH_BOT_VIS=ON
* nuclear strike (deleting build, pod-build, and drake/pod-build), then re-compiling. 

The last one finally got it working :wink: I suspect the problem was that bot2-vis.pc was still floating around in my build folder, triggering the build of `drake_viewer` even though I didn't have bot-vis.

patmarion commented:
Thanks for the report.  Perhaps make clean should rm the build directory?

patmarion commented:
btw, here's my make clean strategy:

rm -rf *
git checkout .
make

:)

You have to be willing to lose local git branches/remotes of your externals.  But sometimes cmake external project wipes things out anyway, so you gotta keep those branches backed up!
